### PR TITLE
Context refactoring

### PR DESCRIPTION
**Pull request by osa1**
Second part of https://github.com/kframework/k/wiki/Design-of-the-new-compiler-parser .

Some things listed in the wiki link may not be possible:
- ~~`Map<String, ASTNode> locations`: Not possible to move to Maude executor, because generated by `DefinitionLoader` which runs in kompile. We have to serialize this field and load in Maude krun.~~
- ~~`Set<Sort> variableTokenSorts`: This is still used, are we sure that this won't break anyting? Or is it a mistake in the wiki page?~~

These are also missing:
- `SetMultiMap<String, Production> tags`: This should be possible, but it's a bit tricky. The problem is that currently it's collecting while productions are parsed to prevent redundant computations. I need to find a way to generate it after productions are added.

### CODE REVIEW

**Comment by dwightguth** [comment]
I thought we were going to serialize the definition though? that should allow locations to be populated in krun.

I am sure regarding variableTokenSorts. This is leftover code from the maude symbolic backend and should be removed.

Please let me know if you have any questions about how you would like to be able to populate tags. I agree it should be possible.

File: kernel/src/main/java/org/kframework/kil/Cell.java [hunk 1]
```diff
@@ -56,6 +56,12 @@
     String endLabel;
 
     /**
+      * Builtin data structure (list, map, set) for the cell.
+      * Set by {@link org.kframework.compile.transformers.Cell2DataStructure}.
+      */
+    public CellDataStructure dataStructure;
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kil/Cell.java]: I meant an Attribute, not a field on this class.

**Comment by osa1** [comment]
Looks like build is failed because of some of the recent commits. I'll rebase this.

File: kernel/src/main/java/org/kframework/kil/Cell.java (hunk 1 above)
Comment (osa1) [inline kernel/src/main/java/org/kframework/kil/Cell.java]: Sorry, what do you mean by an attribute?

Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kil/Cell.java]: I mean look at org.kframework.kil.Attribute and the helper functions
associated with that class in ASTNode.

On Mon, Nov 3, 2014 at 11:20 AM, Ömer Sinan Ağacan <notifications@github.com

**Comment by radumereuta** [comment]
I think this is ok.

**Comment by osa1** [comment]
I had a problem with Cell attributes related with Java backend(stack trace of error site is completely unrelated, but it has to be related with Cell attributes because it's only happening with Cell attributes commit) so removed that commit for now.

**Comment by osa1** [comment]
@dwightguth 

> I am sure regarding variableTokenSorts. This is leftover code from the maude symbolic backend and should be removed.

Removing that broke Java backend. (commit: https://github.com/osa1/k/commit/45eb042227c02f748e76f844383eb40c3022d40a) Should I put that back?

**Comment by dwightguth** [comment]
Per our discussion with Cosmin, I suggest you do the minimum amount of work needed to get this cleaned up and passing on some subset of the tasks we intended to complete, and then merge it in that form before moving on to the data structure refactoring itself.

**Comment by osa1** [comment]
There's this problem with moving cell data structure out of Context: Java backend is creating sorts with hard-coded labels like "generatedTop". Previously cell data structures where queried using cell labels this was working, but since we removed CellLabel -> CellDataStructure map it can't find the data structure and(because now it's searching the data structure using fresh Cell with label "generatedTop") I think this is what's making Java backend to fail.

**Comment by dwightguth** [comment]
Yeah just revert that part for now since we intend to address these issues
primarily later.

**Comment by osa1** [comment]
Finally, tests are passing.

@dwightguth please review.

I didn't make any changes about `tags`, `cellDataStructures` and `variableTokenSorts`.

**Comment by dwightguth** [comment]
Excellent. Go ahead and work with Cosmin on the KORE refactorings. Let me know if you run into situations where you don't have something to work on.

**Pull request merged**